parsec/scheduling.c

```diff
@@ -322,14 +327,58 @@ int __parsec_schedule(parsec_execution_stream_t* es,
     }
 #endif  /* defined(PARSEC_DEBUG_PARANOID) || defined(PARSEC_DEBUG_NOISIER) */
 
-    len = 0;
+#if defined(PARSEC_PAPI_SDE)
+    int len = 0;
     _LIST_ITEM_ITERATOR(task, &task->super, item, {len++; });
     PARSEC_PAPI_SDE_COUNTER_ADD(PARSEC_PAPI_SDE_TASKS_ENABLED, len);
+#endif  /* defined(PARSEC_PAPI_SDE) */
+
     ret = parsec_current_scheduler->module.schedule(es, tasks_ring, distance);
 
     return ret;
 }
 
+/*
+ * Schedule an array of rings of tasks with one entry per virtual process. Each entry contains a ring
+ * of tasks similar to __parsec_schedule. By default this version will save the highest priority
+ * task (assuming the ring is ordered or the first task in the ring otherwise) on the current execution
+ * stream virtual process as the next task to be executed on the current execution stream. Everything
+ * else gets pushed into the execution stream 0 of the corresponding virtual process.
+ */
+int __parsec_schedule_vp(parsec_execution_stream_t* es,
+                         parsec_task_t** task_rings,
+                         int32_t distance)
+{
+    parsec_execution_stream_t* target_es;
+    const parsec_vp_t** vps = (const parsec_vp_t**)es->virtual_process->parsec_context->virtual_processes;
+    int ret = 0;
+
+    for(int vp = 0; vp < es->virtual_process->parsec_context->nb_vp; vp++ ) {
+        parsec_task_t* ring = task_rings[vp];
+        if( NULL == ring ) continue;
+
+        target_es = vps[vp]->execution_streams[0];
+
+        if( vp == es->virtual_process->vp_id ) {
+            if( NULL == es->next_task ) {
```

Comment:
Can this function ever be called from the comm thread? If so, we might lose tasks because the comm thread uses the same `es` as thread 0

Reply:
the comm thread, in fact no threads, should not call this function with that `es`. It can be called with NULL, or with its own copy where the `next_task` field is set to a predefined value.